okhttp/src/main/java/okhttp3/Dispatcher.java

```diff
@@ -173,6 +185,10 @@ synchronized void executed(RealCall call) {
   /** Used by {@code Call#execute} to signal completion. */
   synchronized void finished(Call call) {
     if (!runningSyncCalls.remove(call)) throw new AssertionError("Call wasn't in-flight!");
+
+    if (runningCallsCount() == 0 && idleCallback != null) {
```

Comment:
It’ll make the code a little uglier, but I think it’s probably worthwhile to rearrange this so we call the callback outside of the `synchronized` block. As is if the callback takes a while to run then no calls can be enqueued which is bad.